### PR TITLE
Reimplement action parameters mapping and execution

### LINKED ISSUE #25234: [App] Add custom action to page

### LINKED ISSUE #25235: [App] Map parameters to actions

### PR DESCRIPTION
**Pull request by kulyk**
The core part of #25265 (read for more context). Closes #25234, closes #25235

Reimplements writeback actions parameter mapping and executions code (removed in #25255). The main difference is that we don't use `emitter` objects anymore and don't rely on action button dash cards `click_behavior` for mapping parameters and handling button clicks. Instead, dash cards now have an `action_id` column we can set, and we can re-use the `parameter_mappings` column.

**Highlights**

1. Earlier we extended [`DashboardClickDrill`](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/modes/components/drill/DashboardClickDrill.jsx) to also perform writeback actions. Instead, this PR introduces a new [`ActionClickDrill`](https://github.com/metabase/metabase/blob/new-action-exec/2-new-approach/frontend/src/metabase/modes/components/drill/ActionClickDrill/ActionClickDrill.tsx) that's responsible only for running writeback actions.
2. We're still using the click behavior sidebar for configuring action buttons, although we're not setting the `click_behavior` property onto the dash card `visualization_settings`
3. We're re-using the `ClickMappings` component for configuring action parameters and dashboard parameter mappings. `ClickMappings` only works with the `click_behavior` shape, so this PR introduces [`ActionClickMappings` component](https://github.com/metabase/metabase/blob/new-action-exec/2-new-approach/frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/ActionOptions/ActionClickMappings.tsx) that serves as an adapter between two formats.

### Demo

https://user-images.githubusercontent.com/17258145/188714110-e7b257a1-16be-4353-9c6c-dd1d19d68e99.mp4

### CODE REVIEW

File: frontend/src/metabase/modes/components/drill/ActionClickDrill/ActionClickDrill.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  const parameters: ParameterMappedForActionExecution[] = [];
+  const data = getDataFromClicked(clicked);
+
+  dashcard.parameter_mappings?.forEach?.(mapping => {
+    const parameter = prepareParameter(mapping, { action, data });
+    if (parameter) {
+      parameters.push(parameter);
+    }
+  });
```
Comment (iethree) [inline frontend/src/metabase/modes/components/drill/ActionClickDrill/ActionClickDrill.tsx:26]: This is a little hard to follow setting this to an empty array and conditionally pushing into it later. I know it's a little less efficient, but is this maybe a bit easier to read?


```suggestion
  const data = getDataFromClicked(clicked);

  const parameters: ParameterMappedForActionExecution[] = 
    dashcard.parameter_mappings?.map?.(mapping => (
      prepareParameter(mapping, { action, data })
    )).filter(Boolean);
```

File: frontend/src/metabase/services.js [hunk 2]
```diff
@@ -536,4 +536,7 @@ export const ActionsApi = {
   delete: POST("/api/action/row/delete"),
   bulkUpdate: POST("/api/action/bulk/update/:tableId"),
   bulkDelete: POST("/api/action/bulk/delete/:tableId"),
+  execute: POST(
+    "/api/dashboard/:dashboardId/dashcard/:dashcardId/action/:actionId/execute",
+  ),
```
Comment (iethree) [inline frontend/src/metabase/services.js:541]: @snoe or @metamben what happens if we pass the wrong action id here? i.e. an action id that doesn't match the action id that is actually linked to the dashcard?

**Review by iethree — APPROVED** [review]
Code looks good as far as I can tell. Was not able to test locally because whenever I try to save any dashboard (or app page) I get this error from the backend

```
 ["JdbcSQLException:"
  "Message: Column \"SIZEX\" not found; SQL statement:"
  "INSERT INTO \"REPORT_DASHBOARDCARD\" (\"SIZEX\", \"ACTION_ID\", \"UPDATED_AT\", \"COL\", \"PARAMETER_MAPPINGS\", \"CARD_ID\", \"ENTITY_ID\", \"VISUALIZATION_SETTINGS\", \"DASHBOARD_ID\", \"CREATED_AT\", \"SIZEY\", \"ROW\") VALUES (2, NULL, now(), 0, ?, 72, ?, ?, 32, now(), 2, 0) [42122-197]"
  "SQLState: 42S22"
  "Error Code: 42122"]}
```

**Comment by kulyk** [comment]
> Was not able to test locally because whenever I try to save any dashboard (or app page) I get this error from the backend

@iethree oh, seems like it's caused by #25192; I suppose a rebase on fresh `master` should fix this, I'll take a look

**Comment by kulyk** [comment]
@iethree should be fine now, sorry for the confusion

https://user-images.githubusercontent.com/17258145/188879282-b134d542-abc0-4a91-ba73-94221cb74764.mp4

File: frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/ActionOptions/ActionOptionItem.tsx [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  isSelected,
+  onClick,
+}: ActionOptionProps) {
+  return (
+    <ActionSidebarItem
+      onClick={onClick}
+      isSelected={isSelected}
+      hasDescription={!!description}
+    >
+      <ActionSidebarItemIcon name="bolt" isSelected={isSelected} />
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/ActionOptions/ActionOptionItem.tsx:29]: Good catch, thank you! I checked Figma once again, and we're actually using the "insight" icon there

File: frontend/src/metabase/modes/components/drill/ActionClickDrill/ActionClickDrill.tsx (hunk 1 above)
Comment (kulyk) [reply frontend/src/metabase/modes/components/drill/ActionClickDrill/ActionClickDrill.tsx:26]: Sounds good to me, but apparently a change like that doesn't make TypeScript happy, see:

```typescript
const parameterMappings = dashcard.parameter_mappings || [];

const parameters: ParameterMappedForActionExecution[] = parameterMappings
  .map(mapping => prepareParameter(mapping, { action, data }))
  .filter(Boolean);
```

1. `prepareParameter` can return `undefined`
2. `map` returns `Array<ParameterMappedForActionExecution | undefined>`
3. TypeScript is not convinced that `filter(Boolean)` would make it `Array<ParameterMappedForActionExecution>`

Check out [this StackOverflow question](https://stackoverflow.com/questions/47632622/typescript-and-filter-boolean). I'm worried a fix like that can be more confusing that iterating with `forEach` and conditionally pushing into a list :( Please LMK if you see a better solution here

Comment (kulyk) [reply frontend/src/metabase/modes/components/drill/ActionClickDrill/ActionClickDrill.tsx:26]: @iethree if you don't mind, I'd merge this PR now and we can iterate on the implementation details here in #25257

File: frontend/src/metabase/services.js (hunk 2 above)
Comment (metamben) [reply frontend/src/metabase/services.js:541]: @iethree a 404 happens because no entry in `report_dashboardcard` will be found in the DB with those IDs.
